File: src/go/piperplus/voice_catalog.go
package piperplus

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// VoiceFileInfo describes a file in a voice model package.
type VoiceFileInfo struct {
	RelativePath string `json:"relative_path"`
	SizeBytes    int64  `json:"size_bytes"`
	MD5          string `json:"md5,omitempty"`
}

// VoiceCatalogEntry describes a voice model in the catalog.
type VoiceCatalogEntry struct {
	Key             string          `json:"key"`
	Name            string          `json:"name"`
	LanguageCode    string          `json:"language_code"`
	LanguageFamily  string          `json:"language_family,omitempty"`
	LanguageNative  string          `json:"language_native,omitempty"`
	LanguageEnglish string          `json:"language_english,omitempty"`
	Quality         string          `json:"quality"`
	NumSpeakers     int             `json:"num_speakers"`
	Source          string          `json:"source,omitempty"`
	RepoID          string          `json:"repo_id,omitempty"`
	Files           []VoiceFileInfo `json:"files,omitempty"`
	Aliases         []string        `json:"aliases,omitempty"`
}

// HasAlias reports whether the entry has the given alias.
func (e *VoiceCatalogEntry) HasAlias(alias string) bool {
	for _, a := range e.Aliases {
		if a == alias {
			return true
		}
	}
	return false
}

// OnnxFileName returns the .onnx file name from the entry's file list, or empty string.
func (e *VoiceCatalogEntry) OnnxFileName() string {
	for _, f := range e.Files {
		if filepath.Ext(f.RelativePath) == ".onnx" {
			return filepath.Base(f.RelativePath)
		}
	}
	return ""
}

// embeddedCatalog contains built-in piper-plus voice entries.
var embeddedCatalog = []VoiceCatalogEntry{
	{
		Key: "ja_JP-tsukuyomi-chan-medium", Name: "Tsukuyomi-chan",
		LanguageCode: "ja_JP", LanguageFamily: "ja", LanguageNative: "日本語", LanguageEnglish: "Japanese",
		Quality: "medium", NumSpeakers: 1, Source: "piper-plus", RepoID: "ayousanz/piper-plus-tsukuyomi-chan",
		Files: []VoiceFileInfo{
			{RelativePath: "tsukuyomi-chan.onnx", SizeBytes: 63_000_000},
			{RelativePath: "tsukuyomi-chan.onnx.json", SizeBytes: 15_000},
		},
		Aliases: []string{"tsukuyomi-chan", "tsukuyomi"},
	},
	{
		Key: "ja_JP-css10-6lang-medium", Name: "CSS10 6-Language",
		LanguageCode: "ja_JP", LanguageFamily: "ja", LanguageNative: "日本語", LanguageEnglish: "Japanese",
		Quality: "medium", NumSpeakers: 1, Source: "piper-plus", RepoID: "ayousanz/piper-plus-base",
		Files: []VoiceFileInfo{
			{RelativePath: "multilingual-test-medium.onnx", SizeBytes: 38_000_000},
			{RelativePath: "multilingual-test-medium.onnx.json", SizeBytes: 15_000},
		},
		Aliases: []string{"css10", "css10-6lang"},
	},
}

// FindVoice looks up a voice by exact key, then by alias.
func FindVoice(nameOrAlias string) (*VoiceCatalogEntry, bool) {
	for i := range embeddedCatalog {
		if embeddedCatalog[i].Key == nameOrAlias {
			entry := embeddedCatalog[i]
			return &entry, true
		}
	}
	for i := range embeddedCatalog {
		if embeddedCatalog[i].HasAlias(nameOrAlias) {
			entry := embeddedCatalog[i]
			return &entry, true
		}
	}
	return nil, false
}

// ListVoices returns catalog entries, optionally filtered by language code prefix.
func ListVoices(languageFilter string) []VoiceCatalogEntry {
	var result []VoiceCatalogEntry
	for _, e := range embeddedCatalog {
		if languageFilter == "" || strings.HasPrefix(e.LanguageCode, languageFilter) || e.LanguageFamily == languageFilter {
			result = append(result, e)
		}
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i].LanguageCode != result[j].LanguageCode {
			return result[i].LanguageCode < result[j].LanguageCode
		}
		return result[i].Key < result[j].Key
	})
	return result
}

// LoadExternalCatalog loads a voices.json file.
func LoadExternalCatalog(path string) ([]VoiceCatalogEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var entries []VoiceCatalogEntry
	if err := json.Unmarshal(data, &entries); err != nil {
		return nil, err
	}
	return entries, nil
}

File: src/csharp/PiperPlus.Core/Config/VoiceCatalog.cs
using System.Text.Json;

namespace PiperPlus.Core.Config;

/// <summary>
/// Provides access to the piper-plus voice catalog.
/// Mirrors the C++ <c>loadVoiceCatalog()</c> function in <c>model_manager.cpp</c>.
/// </summary>
public static class VoiceCatalog
{
    // -------------------------------------------------------------------
    // Built-in catalog (object initializers — no JSON parsing at runtime)
    // -------------------------------------------------------------------

    private static readonly VoiceInfo[] BuiltInVoices =
    [
        new VoiceInfo(
            Key: "ja_JP-tsukuyomi-chan-medium",
            Name: "tsukuyomi-chan",
            LanguageCode: "ja_JP",
            LanguageFamily: "ja",
            LanguageNameNative: "日本語",
            LanguageNameEnglish: "Japanese",
            Quality: "medium",
            NumSpeakers: 1,
            Source: "piper-plus",
            RepoId: "ayousanz/piper-plus-tsukuyomi-chan",
            Files:
            [
                new VoiceFileInfo("tsukuyomi-chan-6lang-fp16.onnx", 77594624, ""),
                new VoiceFileInfo("config.json", 3072, ""),
            ],
            Aliases: ["tsukuyomi", "tsukuyomi-chan", "ja-tsukuyomi"],
            Description: "Tsukuyomi-chan Japanese TTS model trained with WavLM discriminator (300 epochs)"),

        new VoiceInfo(
            Key: "ja_JP-css10-6lang-medium",
            Name: "css10-6lang",
            LanguageCode: "ja_JP",
            LanguageFamily: "ja",
            LanguageNameNative: "日本語",
            LanguageNameEnglish: "Japanese",
            Quality: "medium",
            NumSpeakers: 1,
            Source: "piper-plus",
            RepoId: "ayousanz/piper-plus-css10-ja-6lang",
            Files:
            [
                new VoiceFileInfo("css10-ja-6lang-fp16.onnx", 39414515, ""),
                new VoiceFileInfo("config.json", 8966, ""),
            ],
            Aliases: ["css10", "css10-6lang", "css10-ja", "ja-css10"],
            Description: "CSS10 Japanese 6-language TTS model fine-tuned from multilingual base (FP16, 6841 utterances)"),
    ];

    /// <summary>
    /// Returns the embedded piper-plus catalog (no I/O).
    /// </summary>
    public static IReadOnlyList<VoiceInfo> LoadBuiltInCatalog()
        => BuiltInVoices;

    /// <summary>
    /// Loads a voice catalog from an external <c>voices.json</c> file.
    /// The JSON format matches the C++ upstream catalog: a dictionary keyed by voice key.
    /// </summary>
    /// <exception cref="FileNotFoundException">Thrown when <paramref name="path"/> does not exist.</exception>
    /// <exception cref="InvalidOperationException">Thrown when deserialization fails.</exception>
    public static IReadOnlyList<VoiceInfo> LoadFromFile(string path)
    {
        if (!File.Exists(path))
        {
            throw new FileNotFoundException(
                $"Voices catalog file not found: {path}", path);
        }

        using var stream = File.OpenRead(path);
        var dict = JsonSerializer.Deserialize(
            stream,
            VoiceCatalogJsonContext.Default.DictionaryStringVoiceJsonEntry);

        if (dict is null)
        {
            throw new InvalidOperationException(
                $"Failed to deserialize voice catalog from: {path}");
        }

        var result = new List<VoiceInfo>(dict.Count);
        foreach (var (key, entry) in dict)
        {
            result.Add(VoiceJsonConverter.ToVoiceInfo(key, entry));
        }

        return result;
    }

    private static readonly Lazy<IReadOnlyList<VoiceInfo>> s_cachedCatalog =
        new(() => LoadMergedCatalogInternal(null));

    /// <summary>
    /// Returns the built-in catalog merged with an optional external <c>voices.json</c>.
    /// Built-in entries take precedence when keys collide (same semantics as the C++ implementation).
    /// The result is sorted by language code, then by key.
    /// When <paramref name="externalVoicesJsonPath"/> is <c>null</c>, the result is cached
    /// (thread-safe via <see cref="Lazy{T}"/>).
    /// </summary>
    public static IReadOnlyList<VoiceInfo> LoadMergedCatalog(string? externalVoicesJsonPath = null)
    {
        if (externalVoicesJsonPath is null)
            return s_cachedCatalog.Value;

        // External file specified — bypass cache
        return LoadMergedCatalogInternal(externalVoicesJsonPath);
    }

    private static IReadOnlyList<VoiceInfo> LoadMergedCatalogInternal(string? externalVoicesJsonPath)
    {
        var builtIn = LoadBuiltInCatalog();

        if (string.IsNullOrEmpty(externalVoicesJsonPath) || !File.Exists(externalVoicesJsonPath))
        {
            // Nothing to merge — return built-in sorted
            return SortCatalog(builtIn);
        }

        IReadOnlyList<VoiceInfo> external;
        try
        {
            external = LoadFromFile(externalVoicesJsonPath);
        }
        catch (Exception ex)
        {
            // If external file is unreadable, fall back to built-in only
            System.Diagnostics.Debug.WriteLine(
                $"Failed to load external voices catalog: {ex.Message}");
            return SortCatalog(builtIn);
        }

        // Merge: built-in keys win
        var merged = new Dictionary<string, VoiceInfo>(
            builtIn.Count + external.Count, StringComparer.Ordinal);

        foreach (var voice in builtIn)
        {
            merged[voice.Key] = voice;
        }

        foreach (var voice in external)
        {
            merged.TryAdd(voice.Key, voice);
        }

        return SortCatalog(merged.Values);
    }

    // -------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------

    /// <summary>
    /// Sorts voices by language code, then by key (same order as C++ loadVoiceCatalog).
    /// </summary>
    private static IReadOnlyList<VoiceInfo> SortCatalog(IEnumerable<VoiceInfo> voices)
    {
        var list = voices.ToList();
        list.Sort((a, b) =>
        {
            int cmp = string.Compare(a.LanguageCode, b.LanguageCode, StringComparison.Ordinal);
            return cmp != 0 ? cmp : string.Compare(a.Key, b.Key, StringComparison.Ordinal);
        });
        return list;
    }
}
